e2e: add tests to validate omap objects

pkalever opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #
Add E2E tests to validate OMAP data for both CephFS and RBD for all operations such as:

- CreateVolume
- DeleteVolume
- CreateSnapshot
- DeleteSnapshot



## Related issues ##

Fixes: #2834

Signed-off-by: Prasanna Kumar Kalever \<prasanna.kalever@redhat.com\>

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

pkalever commented:
/retest all

pkalever commented:
@Mergifyio refresh

Commit:
e2e: add utility functions

such as:
listCephFSFileSystems()
getCephFSPoolName()
validateOmapCount()
compareStdoutWithCount()

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
e2e: validate omap objects per rbd test case

validate omap count in every testcase right after
validateRBDImageCount()

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
e2e: validate omap objects per cephfs test case

validate omap count in every testcase right after
validateSubvolumeCount()

Fixes: #2834
Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

pkalever commented on e2e/utils.go:
```diff
@@ -82,6 +82,156 @@ var (
+			e2elog.Failf("failed to convert string to int: %s", stdOut)
+		}
+		if res != count {
+			e2elog.Logf("expected omap object count %d, got %d", count, res)
```
I have intentionally used `e2elog.Logf` instead of `e2elog.Failf` in this function in a few places. I will remove them later, I just want the first run to finish without stopping on any single failure.

nixpanic replied:
do you want to move this to a **Draft** in the mean time?

pkalever replied:
Yes, Good idea.

Rakshith-R commented on e2e/cephfs.go:
```diff
@@ -304,12 +304,20 @@ var _ = Describe("cephfs", func() {
 						e2elog.Failf("failed to create application: %v", err)
 					}
 					validateSubvolumeCount(f, 1, fileSystemName, subvolumegroup)
+					pool, err := getCephFSPoolName(f, fileSystemName)
+					if err == nil {
+						e2elog.Logf("poolName: %s", pool)
+						validateOmapCount(f, 1, "cephfs", pool, "volumes")
+					}
```
Get the pool name only once, before all the `By("..` ?

Rakshith-R commented on e2e/utils.go:
```diff
@@ -49,6 +49,9 @@ const (
 	defaultNs     = "default"
 	defaultSCName = ""
 
+	rbdType    = "rbd"
+	cephfsType = "cephfs"
+
```
Use these constants in both e2e/cephfs.go and e2e/rbd.go.

Madhu-1 commented on e2e/utils.go:
```diff
@@ -82,6 +85,133 @@ var (
+		return v.MetadataPool, nil
+	}
+
+	return "", os.ErrNotExist
```
`os.ErrNotExist` is not a proper error message for pools as its for files. use fmt.Errorf here?

Madhu-1 commented on e2e/utils.go:
```diff
@@ -82,6 +85,133 @@ var (
+}
+
+// getCephFSPoolName get CephFS pool name from filesystem name.
+func getCephFSPoolName(f *framework.Framework, filesystem string) (string, error) {
```
getCephFSPoolName to `getCephFSMetadataPoolName`

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -304,12 +304,20 @@ var _ = Describe("cephfs", func() {
 						e2elog.Failf("failed to create application: %v", err)
 					}
 					validateSubvolumeCount(f, 1, fileSystemName, subvolumegroup)
+					pool, err := getCephFSPoolName(f, fileSystemName)
```
```suggestion
metadadataPool, err := getCephFSMetadataPoolName(f, fileSystemName)
```

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -304,12 +304,20 @@ var _ = Describe("cephfs", func() {
 					}
 					validateSubvolumeCount(f, 0, fileSystemName, subvolumegroup)
+					if pool != "" {
+						validateOmapCount(f, 0, "cephfs", pool, "volumes")
```
define consts for `cephfs` and `volumes`

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -351,12 +359,20 @@ var _ = Describe("cephfs", func() {
 					}
 					validateSubvolumeCount(f, 1, fileSystemName, subvolumegroup)
+					pool, err := getCephFSPoolName(f, fileSystemName)
+					if err == nil {
```
Hope the check will be inverted like below

```suggestion
pool, err := getCephFSPoolName(f, fileSystemName)
if err != nil {
e2elog.Failf(err)
}
```

Madhu-1 commented on e2e/utils.go:
```diff
@@ -82,6 +85,133 @@ var (
+type cephfsFilesystem struct {
+	Name           string   `json:"name"`
+	MetadataPool   string   `json:"metadata_pool"`
+	MetadataPoolID int      `json:"metadata_pool_id"`
+	DataPoolIDs    []int    `json:"data_pool_ids"`
+	DataPools      []string `json:"data_pools"`
```
Remove these fields as we dont use them.

Madhu-1 commented on e2e/utils.go:
```diff
@@ -82,6 +85,133 @@ var (
+		}
+		stdOut, stdErr, err = execCommandInToolBoxPod(f, cmd1, rookNamespace)
+		if err != nil {
+			if !strings.Contains(err.Error(), "command terminated with exit code 1") {
```
why string comparison is required here?

pkalever replied:
for example in command:
`rados ls --pool=%s --namespace csi | grep -v default | grep -c ^csi.volume.`
if the `rados ls --pool=%s --namespace csi` results nothing then later part of the command `| grep -v default | grep -c ^csi.volume.` will result in `"command terminated with exit code 1"`

```
sh-4.4$ rados ls --pool=replicapool --namespace csi
sh-4.4$ echo $?
0
sh-4.4$ rados ls --pool=replicapool --namespace csi | grep -v default | grep -c ^csi.volume.
0
sh-4.4$ echo $?
1

```

Madhu-1 commented on e2e/utils.go:
```diff
@@ -82,6 +85,133 @@ var (
+		default:
+			e2elog.Failf("unknown driver type: %s", driver)
+		}
+		stdOut, stdErr, err = execCommandInToolBoxPod(f, cmd1, rookNamespace)
```
add error check for stdErr here and in other places also

Madhu-1 commented on e2e/utils.go:
```diff
@@ -82,6 +85,133 @@ var (
+// Works with Cephfs and RBD drivers and mode can be snaps or volumes.
+func validateOmapCount(f *framework.Framework, count int, driver, pool, mode string) {
+	var (
+		cmd1   string
+		cmd2   string
```
```suggestion
radosLsCmd   string
radosLsKeysCmd   string
```
we can use better naming for the cmds?

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

pkalever commented:
@Madhu-1 @nixpanic as expected we caught the OMAP leaks 
```
Apr 26 16:29:26.446: INFO: expected omap object count 1, got 3
Apr 26 16:29:26.451: INFO: ExecWithOptions {Command:[/bin/sh -c rados listomapkeys csi.volumes.default --pool=replicapool | wc -l] Namespace:rook-ceph PodName:rook-ceph-tools-6f56cdd85d-2vrm5 ContainerName:rook-ceph-tools Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
Apr 26 16:29:26.451: INFO: >>> kubeConfig: /root/.kube/config
Apr 26 16:29:26.451: INFO: ExecWithOptions: Clientset creation
Apr 26 16:29:26.451: INFO: ExecWithOptions: execute(POST https://192.168.39.62:8443/api/v1/namespaces/rook-ceph/pods/rook-ceph-tools-6f56cdd85d-2vrm5/exec?command=%2Fbin%2Fsh&command=-c&command=rados+listomapkeys+csi.volumes.default+--pool%3Dreplicapool+%7C+wc+-l&container=rook-ceph-tools&container=rook-ceph-tools&stderr=true&stdout=true %!s(MISSING))
Apr 26 16:29:26.830: INFO: expected omap object count 1, got 3
```

PTAL

Madhu-1 commented:
> @Madhu-1 @nixpanic as expected we caught the OMAP leaks
> 
> ```
> Apr 26 16:29:26.446: INFO: expected omap object count 1, got 3
> Apr 26 16:29:26.451: INFO: ExecWithOptions {Command:[/bin/sh -c rados listomapkeys csi.volumes.default --pool=replicapool | wc -l] Namespace:rook-ceph PodName:rook-ceph-tools-6f56cdd85d-2vrm5 ContainerName:rook-ceph-tools Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
> Apr 26 16:29:26.451: INFO: >>> kubeConfig: /root/.kube/config
> Apr 26 16:29:26.451: INFO: ExecWithOptions: Clientset creation
> Apr 26 16:29:26.451: INFO: ExecWithOptions: execute(POST https://192.168.39.62:8443/api/v1/namespaces/rook-ceph/pods/rook-ceph-tools-6f56cdd85d-2vrm5/exec?command=%2Fbin%2Fsh&command=-c&command=rados+listomapkeys+csi.volumes.default+--pool%3Dreplicapool+%7C+wc+-l&container=rook-ceph-tools&container=rook-ceph-tools&stderr=true&stdout=true %!s(MISSING))
> Apr 26 16:29:26.830: INFO: expected omap object count 1, got 3
> ```
> 
> PTAL

@pkalever great!. can you please also check where the leak is happening?

Madhu-1 commented on e2e/utils.go:
```diff
@@ -82,6 +88,112 @@ var (
+		e2elog.Failf("failed to convert string to int: %s", stdOut)
+	}
+	if res != count {
+		e2elog.Logf("expected omap object count %d, got %d", count, res)
```
log stdOut here for better debugging.

pkalever replied:
StdOut is just a digit in its string form, which we anyway log after converting it to int as `res`

Madhu-1 replied:
To help to debug we need a way to log the output from the `rados` command.

Madhu-1 replied:
@pkalever can you please check this one? this is really helpful for debugging

pkalever commented:
/retest all

Madhu-1 commented on e2e/utils.go:
```diff
@@ -82,6 +88,112 @@ var (
+}
+
+// getCephFSMetadataPoolName get CephFS pool name from filesystem name.
+func getCephFSMetadataPoolName(f *framework.Framework, filesystem string, metadataPool *string) {
```
why metadataPool as a pointer? just return string from here

Madhu-1 commented on e2e/utils.go:
```diff
@@ -82,6 +88,112 @@ var (
+
+		return
+	}
+	e2elog.Failf("metadata pool name not found for filesystem: %s", filesystem)
```
lets not fail in internal function, return from here and fail in the caller

pkalever replied:
Returning error will require huge changes to e2e, as getCephFSMetadataPoolName() is getting called before all the `By() {}`  all the `err` variables in the `By() { }` need modified, And the linter is throwing shadow declaration errors.

I saw other functions using a similar style too.

pkalever replied:
Now you have asked I will return error but just use a different variable for err may be `err0`

Madhu-1 commented on e2e/utils.go:
```diff
@@ -82,6 +88,112 @@ var (
+	}
+	stdOut, stdErr, err := execCommandInToolBoxPod(f, radosLsCmd, rookNamespace)
+	if err != nil || stdErr != "" {
+		if !strings.Contains(err.Error(), "command terminated with exit code 1") {
```
define string for `"command terminated with exit code 1"` and use it?

Madhu-1 reviewed (commented):
changes LGTM, small suggestions

Madhu-1 commented:
Is @pkalever PR still in the draft?

Commit:
e2e: workaround fix for omap leaks

2 omap objects are getting leaked in the e2e tests, this change is to
workaround them for now.

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

pkalever commented:
> Is @pkalever PR still in the draft?

Not anymore! Thanks!

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -2595,10 +2662,16 @@ var _ = Describe("RBD", func() {
 					e2elog.Failf("failed to check csi journal in pool: %v", err)
 				}
 
+				err = deleteJournalInfoInPool(f, pvc, "replicapool")
+				if err != nil {
+					e2elog.Failf(" failed to delete omap data: %v", err)
+				}
```
@pkalever can you explain more about this one? why this is required in E2E.

pkalever replied:
@Madhu-1 The omap tests had caught 2 omap leaks, at each place I used `deleteJournalInfoInPool` there is 1 leak. The description of `deleteJournalInfoInPool` says it clean the omaps.

Is there any other way to clean the omaps here?

Madhu-1 replied:
We should identify why omap is getting leaked in actual code instead of cleaning it in E2E. if we don't do that we might have leaks in the real environment isn't it?

pkalever replied:
I thought the e2e is not cleaning it properly. I didn't understand the e2e test much. Can you please check on it once?

Madhu-1 replied:
we are just validating the omap created by actual cephcsi deployment for each pvc and snapshots in E2E.
it's not a problem with the E2E, it should be a problem with cephcsi itself. can you please open an issue to track this one? once we fix this one we need to remove the workaround.

pkalever replied:
Got it, sure Madhu!

pkalever replied:
Also I shall make it clear in the commit description that this is a workaround fix only.

Rakshith-R replied:
```suggestion
				err = deleteJournalInfoInPool(f, pvc, "replicapool")
				if err != nil {
					e2elog.Failf(" failed to delete omap data: %v", err)
				}
```
@pkalever even if this is a workaround, should it not be after `deletePVCAndApp()` ?

Madhu-1 replied:
Yes correct I missed this one, @pkalever can you please check.

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -254,6 +254,12 @@ var _ = Describe("cephfs", func() {
 			appEphemeralPath := cephFSExamplePath + "pod-ephemeral.yaml"
 			pvcRWOPPath := cephFSExamplePath + "pvc-rwop.yaml"
+
+			metadataPool, err0 := getCephFSMetadataPoolName(f, fileSystemName)
```
```suggestion
var metadataPool string
metadataPool, err = getCephFSMetadataPoolName(f, fileSystemName)
if err != nil {
	e2elog.Failf("failed getting cephFS metadata pool name: %v", err)
}
```

pkalever replied:
This will lead to shadow declaration of `err` in everywhere!!

Madhu-1 replied:
have you tried calling this function in Context outside of It at line 245? it should work, can you please check

pkalever replied:
No! let me give it a try.

pkalever replied:
Done now

pkalever replied:
This seem to panic:
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.23/detail/mini-e2e_k8s-1.23/1505/pipeline

Madhu-1 commented on e2e/utils.go:
```diff
@@ -82,6 +90,150 @@ var (
+			e2elog.Failf("failed to execute rados command '%s' : err=%v stdErr=%s", radosLsCmdFilter, err, stdErr)
+		}
+	}
+	var err0 error
+	err = compareStdoutWithCount(stdOut, count)
+	if err != nil {
+		if !strings.Contains(err.Error(), "expected omap object count") {
+			e2elog.Failf("%v", err)
+		}
+		stdOut, stdErr, err0 = execCommandInToolBoxPod(f, radosLsCmd, rookNamespace)
+		if err0 == nil {
+			e2elog.Logf("additional debug info: rados ls command output: %s, stdErr: %s, err: %v",
+				stdOut, stdErr, err0)
+		}
+		e2elog.Failf("%v", err)
+	}
```
```suggestion
err = compareStdoutWithCount(stdOut, count)
if err != nil {
	e2elog.Failf("%v. command executed %s and command output is %s", err,radosLsCmd,stdOut)
}
```
can be simplified as above?

pkalever replied:
No, we need to execute the command again, it was you who asked for `rados` command output.

Madhu-1 replied:
got it this is without filter to list all the omap entires. should be fine :+1:

Madhu-1 commented on e2e/utils.go:
```diff
@@ -82,6 +90,150 @@ var (
+			e2elog.Failf("failed to execute rados command '%s' : err=%v stdErr=%s", radosLsKeysCmdFilter, err, stdErr)
+		}
+	}
+	err = compareStdoutWithCount(stdOut, count)
+	if err != nil {
+		if !strings.Contains(err.Error(), "expected omap object count") {
+			e2elog.Failf("%v", err)
+		}
+		stdOut, stdErr, err0 = execCommandInToolBoxPod(f, radosLsKeysCmd, rookNamespace)
+		if err0 == nil {
+			e2elog.Logf("additional debug info: rados ls keys command output: %s, stdErr: %s, err: %v",
+				stdOut, stdErr, err0)
+		}
+		e2elog.Failf("%v", err)
+	}
```
same here as well.

Madhu-1 reviewed (commented):
minor nits, LGTM

Madhu-1 commented on e2e/utils.go:
```diff
@@ -82,6 +90,150 @@ var (
+			e2elog.Failf("failed to execute rados command '%s' : err=%v stdErr=%s", radosLsCmdFilter, err, stdErr)
+		}
+	}
+	var err0 error
```
we dont commonly use `err0` variable. this can be renamed to something else like `cmdErr` etc.

pkalever replied:
Sure

pkalever replied:
@Madhu-1 Done now

pkalever commented:
/retest all

pkalever commented:
@nixpanic @Rakshith-R this PR is waiting for your acks please take a look.

Rakshith-R commented:
> @nixpanic @Rakshith-R this PR is waiting for your acks please take a look.

@pkalever I added a comment here, can PTAL?
https://github.com/ceph/ceph-csi/pull/3041/files#r865771287

pkalever commented:
> > @nixpanic @Rakshith-R this PR is waiting for your acks please take a look.
> 
> @pkalever I added a comment here, can PTAL? https://github.com/ceph/ceph-csi/pull/3041/files#r865771287

@Rakshith-R addressed now, sorry I missed it earlier. Thanks!

nixpanic commented on e2e/utils.go:
```diff
@@ -82,6 +90,150 @@ var (
+
+// validateOmapCount validates no of OMAP entries on the given pool.
+// Works with Cephfs and RBD drivers and mode can be snapsType or volumesType.
+// nolint:cyclop // TODO: reduce complexity
```
this is a new function, make sure to not start with such a TODO.

The nested switches are really horrible, there should be a cleaner way of doing this...

How about:

```go
type radosListCommand struct {
  volumeMode string
  driverType string
  radosLsCmd, radosLsCmdFilter string
  radosLsKeysCmd, radosLsKeysCmdFilter string
}

radosListCommands := []radosListCommand{
  radosListCommand{
    volumeMode: volumesType,
    driverType: cephfsType,
    ....
  },
  radosListCommand{
    volumeMode: volumesType,
    driverType: rbdType,
    ....
  },
  ...
}
```

And then loop through the radosListCommands list, and pick the instance that matches `volumeMode` and `driverType`?

That might even reduce complexity enough to be able to drop the `// nolint`.

nixpanic commented on e2e/rbd.go:
```diff
@@ -2666,6 +2666,10 @@ var _ = Describe("RBD", func() {
 				if err != nil {
 					e2elog.Failf("failed to delete PVC and application: %v", err)
 				}
+				err = deleteJournalInfoInPool(f, pvc, "replicapool")
```
can these _leaks_ be fixed otherwise? do they also leak in normal/production environments?

pkalever replied:
I will raise a new issue for this, let's take it there.

nixpanic replied:
Did you do this? Please comment here with links to the issue(s).

nixpanic replied:
seems to be #3111

nixpanic reviewed (changes requested):
nothing too serious, mostly cleanup needed

pkalever commented:
@Madhu-1 @Rakshith-R @nixpanic PTAL

Madhu-1 reviewed (dismissed):
LGTM

pkalever commented:
@Mergify rebase

pkalever commented:
> > @nixpanic @Rakshith-R this PR is waiting for your acks please take a look.
> 
> @pkalever I added a comment here, can PTAL? https://github.com/ceph/ceph-csi/pull/3041/files#r865771287
@Rakshith-R Moving

pkalever commented on e2e/rbd.go:
```diff
@@ -2768,6 +2772,10 @@ var _ = Describe("RBD", func() {
+					err = deleteJournalInfoInPool(f, pvc, "replicapool")
+					if err != nil {
+						e2elog.Failf(" failed to delete omap data: %v", err)
+					}
```
@Rakshith-R @Madhu-1 Moving deleteJournalInfoInPool() after deletePVCAndApp() causes CI failure

```
May  6 12:26:20.293: FAIL:  failed to delete omap data: failed to get pvc: persistentvolumeclaims "rbd-pvc" not found
```

see: https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e_k8s-1.21/runs/3655/nodes/93/log/?start=0

Madhu-1 replied:
@pkalever can you remove this one I want to see what exactly is left behind? if we are not able to find what is the problem we can add it back. just want to give it a try with logs once.

Madhu-1 replied:
@pkalever do we have logs for this one?

Madhu-1 replied:
Do we have any issue opened for this one?

pkalever replied:
Not yet Madhu, I will surely open one by EOD

pkalever replied:
@Madhu-1 opened this one
https://github.com/ceph/ceph-csi/issues/3111

pkalever commented:
@Madhu-1 @nixpanic @Rakshith-R PTAL

pkalever commented:
@Mergifyio rebase

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

nixpanic commented:
This can be rebased once #3005 has been merged

nixpanic commented:
ah, hoped that adding `DNM` would prevent mergify from rebasing automatically... I guess I was too late

nixpanic commented:
@mergifyio rebase
